### PR TITLE
Update tests (using Iris dataset) to use new API

### PR DESCRIPTION
**Pull request by abgoswam**
- Update a couple of tests (using Iris dataset) to use the new API.

### CODE REVIEW

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 1]
```diff
@@ -3,33 +3,45 @@
...
-using TextLoader = Microsoft.ML.Legacy.Data.TextLoader;
 
 namespace Microsoft.ML.Scenarios
 {
-#pragma warning disable 612, 618
     public partial class ScenariosTests
     {
         [Fact]
-        public void TrainAndPredictIrisModelTest()
+        public void New_TrainAndPredictIrisModelTest()
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: >New_ [](start = 20, length = 4)

I believe "New" prefix was used because we had two APIs but now the old API is gone and there is just one API, do you feel we still need to prefix test with "New"?

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 2]
```diff
@@ -3,33 +3,45 @@
...
+                    new TextLoader.Column("SepalLength", DataKind.R4, 1),
+                    new TextLoader.Column("SepalWidth", DataKind.R4, 2),
+                    new TextLoader.Column("PetalLength", DataKind.R4, 3),
+                    new TextLoader.Column("PetalWidth", DataKind.R4, 4)
+                }
+            );
+
+            var pipe = mlContext.Transforms.Concatenate("Features", "SepalLength", "SepalWidth", "PetalLength", "PetalWidth")
+                            .Append(mlContext.Transforms.Normalize("Features"))
+                            .AppendCacheCheckpoint(mlContext)
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: Why are we adding normalizer transform? just curious.

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 3]
```diff
@@ -3,33 +3,45 @@
...
+                            .Append(mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent("Label", "Features", advancedSettings: s => s.NumThreads = 1));
+
+            // Read training and test data sets
+            string dataPath = GetDataPath(TestDatasets.iris.trainFilename);
+            string testDataPath = dataPath;
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: >string testDataPath = dataPath; [](start = 12, length = 31)

Why is this needed? We are using prediction engine to test.

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 4]
```diff
@@ -3,33 +3,45 @@
...
+            var trainData = reader.Read(dataPath);
+            var testData = reader.Read(testDataPath);
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: Why is this needed?

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 1 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: I added it during development when I had both tests in this file.  Forgot yo drop the 'New_' prefix after deleting the old test.  Will fix.

---
In reply to: [245111276](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245111276) [](ancestors = 245111276)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 5]
```diff
@@ -65,53 +77,20 @@ public void TrainAndPredictIrisModelTest()
...
-            Assert.Equal(48, matrix["1", "1"]);
-            Assert.Equal(2, matrix[1, 2]);
-            Assert.Equal(2, matrix["1", "2"]);
-
-            Assert.Equal(0, matrix[2, 0]);
-            Assert.Equal(0, matrix["2", "0"]);
-            Assert.Equal(1, matrix[2, 1]);
-            Assert.Equal(1, matrix["2", "1"]);
-            Assert.Equal(49, matrix[2, 2]);
-            Assert.Equal(49, matrix["2", "2"]);
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: Do we not have confusion matrix in metrics?

File: test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj [hunk 6]
```diff
@@ -22,7 +22,6 @@
     <ProjectReference Include="..\..\src\Microsoft.ML.StaticPipe\Microsoft.ML.StaticPipe.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.TensorFlow.StaticPipe\Microsoft.ML.TensorFlow.StaticPipe.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.TensorFlow\Microsoft.ML.TensorFlow.csproj" />
-    <ProjectReference Include="..\..\src\Microsoft.ML.Legacy\Microsoft.ML.Legacy.csproj" />
```
Comment (codemzs) [inline test/Microsoft.ML.Tests/Microsoft.ML.Tests.csproj]: ><ProjectReference Include="..\..\src\Microsoft.ML.Legacy\Microsoft.ML.Legacy.csproj" /> [](start = 4, length = 87)

Nice!!

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 2 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: I think in the new API  the paradigm is to be explicit  about whether are using  Normalizers  in the pipeline .  

(In the old Legacy API normalizers used to be auto-inserted based on the algorithm)

Hence decided to be explicit about normalizers.  Please see #1604  which talks about  removing 'smarts' (i.e.  no auto-normalization, no auto-caching and no auto-calibration) in the new API


---
In reply to: [245111952](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245111952) [](ancestors = 245111952)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 5 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: I do not see it inside `metrics`.  Who would be the best person to confirm  whether we support `ConfusionMatrix` in the new API ?

If we do, is there an example I can follow ? I do not see it in the CookBook either

---
In reply to: [245114114](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245114114) [](ancestors = 245114114)

**Comment by Ivanidzo4ka** [comment]
@abgoswam Can we just delete this files instead of modifying them? We have test which do evaluation, we have tests which do predictions (with new API), what's the point of having this ones?

**Comment by codemzs** [comment]
@Ivanidzo4ka We did that and it results in code coverage drop. Check out my PR that removes these files. It’s possible removing some tests might be ok but which ones will require tests by tests code coverage analysis. It might just be easier to convert all the test to new API. Even though these tests are doing training, evaluation, etc but they may be using different transforms, trainers, settings, etc.

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs [hunk 7]
```diff
@@ -65,53 +77,20 @@ public void TrainAndPredictIrisModelTest()
...
-            evaluator.OutputTopKAcc = 3;
-            ClassificationMetrics metrics = evaluator.Evaluate(model, testData);
+            // Evaluate the trained pipeline
+            var predicted = trainedModel.Transform(testData);
+            var metrics = mlContext.MulticlassClassification.Evaluate(predicted, topK:3);
 
             Assert.Equal(.98, metrics.AccuracyMacro);
             Assert.Equal(.98, metrics.AccuracyMicro, 2);
-            Assert.Equal(.06, metrics.LogLoss, 2);
+            Assert.InRange(metrics.LogLoss, .05, .06);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: Could we use Assert.Equal(…, …, 2)?

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs [hunk 8]
```diff
@@ -3,126 +3,100 @@
...
-            evaluator.OutputTopKAcc = 3;
-            ClassificationMetrics metrics = evaluator.Evaluate(model, testData);
+            // Evaluate the trained pipeline
+            var predicted = trainedModel.Transform(testData);
+            var metrics = mlContext.MulticlassClassification.Evaluate(predicted, topK: 3);
 
             Assert.Equal(.98, metrics.AccuracyMacro);
             Assert.Equal(.98, metrics.AccuracyMicro, 2);
-            Assert.Equal(.06, metrics.LogLoss, 2);
+            Assert.InRange(metrics.LogLoss, .05, .06);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs]: Assert.Equal(metrics.Logloss, 0.055, 3) maybe?

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs [hunk 9]
```diff
@@ -139,9 +113,17 @@ public class IrisDataWithStringLabel
             [LoadColumn(3)]
             public float PetalWidth;
 
-            [LoadColumn(4), ColumnName("Label")]
+            [LoadColumn(4)]
             public string IrisPlantType;
         }
+
+        public class IrisPredictionWithStringLabel
```
Comment (wschin) [inline test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs]: Are those data structures used outside this file?

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 7 above)
Comment (codemzs) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: I will also prefer Equal

**Comment by Ivanidzo4ka** [comment]
I've run it with tests and without tests, and only dll for which code coverage went down is Legacy, which is expected. Everything else either state same, or went up by 0.01% :)


---
In reply to: [451270185](https://github.com/dotnet/machinelearning/pull/2008#issuecomment-451270185) [](ancestors = 451270185)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 5 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: Created issue #2009  to bring back support for `ConfusionMatrix` in the new API 

---
In reply to: [245117612](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245117612) [](ancestors = 245117612,245114114)

**Comment by codemzs** [comment]
I have explained to you in person the issue with running code coverage per test in the context of parallelizing test conversion effort. Say you have two tests, each assigned to two devs, these two tests are exercising same code paths, if both devs in parallel run code coverage for the test assigned to them then both will end up removing both tests because of this race condition and then we lose code coverage. I do not see this a safe and quick way to prevent regression in code coverage due to the removal Legacy API binary. The fastest and safest way is to convert all the test to the new API. Correct me if I'm wrong.

---
In reply to: [451278660](https://github.com/dotnet/machinelearning/pull/2008#issuecomment-451278660) [](ancestors = 451278660,451270185)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs (hunk 8 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationWithStringLabelTests.cs]: fixed

---
In reply to: [245133430](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245133430) [](ancestors = 245133430)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 1 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: fixed

---
In reply to: [245113327](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245113327) [](ancestors = 245113327,245111276)

File: test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs (hunk 7 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/Scenarios/IrisPlantClassificationTests.cs]: fixed

---
In reply to: [245133023](https://github.com/dotnet/machinelearning/pull/2008#discussion_r245133023) [](ancestors = 245133023)
